reframe/core/schedulers/slurm.py

```diff
@@ -209,21 +212,22 @@ def emit_preamble(self, job):
                 )
             )
 
-        for opt in job.sched_access:
-            if not opt.strip().startswith(('-C', '--constraint')):
-                preamble.append('%s %s' % (self._prefix, opt))
-
-        # To avoid overriding a constraint that's passed into `sched_access`,
-        # we AND it with the `--constraint` option passed either in `options`
-        # or in `cli_options`
         constraints = []
         constraint_parser = ArgumentParser()
         constraint_parser.add_argument('-C', '--constraint')
-        parsed_options, _ = constraint_parser.parse_known_args(
-            job.sched_access
-        )
-        if parsed_options.constraint:
-            constraints.append(parsed_options.constraint.strip())
+        if not self._sched_access_in_submit:
+            for opt in job.sched_access:
+                if not opt.strip().startswith(('-C', '--constraint')):
+                    preamble.append('%s %s' % (self._prefix, opt))
+
+            # To avoid overriding a constraint that's passed into
+            # `sched_access`, we AND it with the `--constraint` option
+            # passed either in `options` or in `cli_options`
+            parsed_options, _ = constraint_parser.parse_known_args(
+                job.sched_access
+            )
+            if parsed_options.constraint:
+                constraints.append(parsed_options.constraint.strip())
```

Comment:
I think that we should do the same thing with constraints in both cases.

Reply:
But in the case the `-C c1` is passed in with `sbatch`and then they are joined in the script `-C c1&c2` the option in the script will probably be ignored completely, right?
So are you suggesting that we join them in the submission command as well? Then the behaviour would be that we only add the constraint options from the test in the job, while passing the rest in the script?

Reply:
Currently, `sched_access` is combined with the last constraint passed as an option, either from the script or from the command-line. The rationale behind that was that a constraint passed either in the test or in the command line would completely override the ReFrame partition definition if we didn't do that. But put aside this rationale, passing the `sched_access` from the command line should be completely equivalent to passing them through the script, so this logic must be preserved.

Reply:
I think additional constraints should be combined with `sched_access` and passed from the command line in this case, because this is the equivalent of what we are doing now.

Reply:
The logic becomes a bit complicated to properly support this, so I suggest postponing this to 4.7 to properly test. Maybe we should change the logic and stop treating constraints specially and pass the responsibility of combining the constraints to the test.

Reply:
I compacted the constraint code in the Slurm backend and change its logic to fit also the case where `sched_access_in_submit=True`. What I'm practically doing is to "move out" the constraint options from either the `job.options` or the `job.cli_options` to the `sched_access` and then we can continue as with the rest of the schedulers.